service/controller/v2/resource/instance/create.go

```diff
@@ -15,11 +18,91 @@ func (r *Resource) EnsureCreated(ctx context.Context, obj interface{}) error {
 		return microerror.Mask(err)
 	}
 
-	r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("EnsureCreated called for cluster ID '%s'", key.ClusterID(customObject)))
+	// Find the next instance ID and instance name we want to trigger the update
+	// for. Instance names look something like the following example.
+	//
+	//     0gjpt-worker-000004
+	//
+	// Instance IDs are simple non negative integers.
+	//
+	//     4
+	//
+	var instanceID string
+	var instanceName string
+	{
+		r.logger.LogCtx(ctx, "level", "debug", "message", "looking for the next instance to be updated")
 
-	// TODO list all instances
-	// TODO find the first instance not having the latest scale set model applied
-	// TODO trigger update for found instance
+		c, err := r.getVMsClient()
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		g := key.ResourceGroupName(customObject)
+		s := key.WorkerVMSSName(customObject)
+		result, err := c.List(ctx, g, s, "", "", "")
+		if IsScaleSetNotFound(err) {
+			r.logger.LogCtx(ctx, "level", "debug", "message", "did not find the scale set")
+			resourcecanceledcontext.SetCanceled(ctx)
+			r.logger.LogCtx(ctx, "level", "debug", "message", "canceling resource for custom object")
+
+			return nil
+		} else if err != nil {
+			return microerror.Mask(err)
+		}
+
+		for _, v := range result.Values() {
```

Comment:
Are values guaranteed to be returned in the same order? I'd sort them first by name to make sure we update only one instance at a time.

Reply:
What problem do we have when not sorting? I do not see it.

Reply:
So there can be more than single instance rolling at a time if rolling takes longer than resync period if I understand it correctly.

Reply:
In this case we see that the provisioning state is not final and cancel reconciliation.

Reply:
So we never update more than one per reconciliation loop. We can batch this in the future when we want to but I would like to do this separately as soon as we have to. For now I would like to just make it work reliably.

Reply:
I understand when `state is not final` then upgrade is in progress.

Let's say on the first loop we receive instances like

```
inst1 not_latest_model final_state
inst2 not_latest_model final_state
```

So we schedule upgrade for `inst1` because it's the first instance with not latest model in final state.

Second loop we receive list like:

```
inst1 not_latest_model not_final_state
inst2 not_latest_model final_state
```

So we don't schedule upgrade because first instance with not latest model is in not final state.

Third loop we receive list like (order not guaranteed):

```
inst2 not_latest_model final_state
inst1 not_latest_model not_final_state
```

So we schedule upgrade for `inst2` because it's the first instance with not latest model in final state.

- - -

Therefore I think we should either sort this list of check if there is *any* instance in not final state.

Reply:
I added another loop for checking if any instance is in a non final provisioning state. Please have another look if this is cool for you.